Update GTK+ to version 3.14.15

nacho opened the pull request:
Based on the patches from Mikkel Kruse Johnsen.

See that the CSS demo is currently disabled since the CssProvider
is not properly picked up for some reason. We will need to research
a bit on this but for now it is good enough doing it without.

monojenkins commented:
Hello! I'm the build bot for the Mono project.

I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done.

Contributors can ignore this message.

monojenkins commented:
Hello! I'm the build bot for the Mono project.

I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done.

Contributors can ignore this message.

knocte commented on gtk-sharp.sln:
```diff
@@ -1,256 +1,256 @@
-﻿
```
Why is this file changed?

knocte commented on generator/ObjectBase.cs:
```diff
@@ -183,7 +183,7 @@ public override string FromNative (string var, bool owned)
 				* <signal> and <virtual_method> elements do not match the struct layout.
 				*/
-				return (is_interface || this.ParserVersion >= 2) && class_fields_valid;
+				return (is_interface || this.ParserVersion >= 2) && (class_fields_valid || class_struct_name == "GtkWidgetClass");
```
Can you explain the rationale for this 1liner?

knocte commented on sources/Makefile.am:
```diff
@@ -1,26 +1,32 @@
 	patch -p0 gtk+-$(TARGET_GTK_VERSION)/gtk/gtktreemodel.c < gtk_tree_model_signal_fix.patch
+	patch -p0 gtk+-$(TARGET_GTK_VERSION)/gtk/gtkselectionprivate.h < gtkselectionprivate-space.patch
+	patch -p0 gtk+-$(TARGET_GTK_VERSION)/gtk/gtktextattributes.h < gtktextattributes-gi-scanner.patch
+	patch -p0 gtk+-$(TARGET_GTK_VERSION)/gtk/gtktextview.h < gtk_text_view_public.patch
```
What are these 3 new patches?

knocte commented:
If this code is based on another PR, please put the link to it, and explain the adjustments you made over his work.
Thanks

nacho commented:
Here we go. Please let me know if you need me to provide more info.

knocte commented:
> Based on the patches from Mikkel Kruse Johnsen. See the fork:
> https://github.com/openmedicus/gtk-sharp/

Prefer that you include the link to his earlier PR, to see previous conversations about this upgrade.

> There are also 3 new patches that need to be applied due to the
scanner hanging or giving an error when parsing the source code.
> Two of those patches are already in GTK upstream see:

Thanks for providing info about this! And URLs of upstream patches. But I only see 2 URLs, not 3.

> See that there are also some generator problems with GtkWidgetClass.

Meh, his commit doesn't delve into what those generator problems are. It could be great if you could give more insight about this before we merge this workaround.

nacho committed:
Update GTK+ to version 3.14.15

Based on the patches from Mikkel Kruse Johnsen. See the PR:
https://github.com/mono/gtk-sharp/pull/128
While Mikkel has patches on his fork to update step by step
to GTK 3.20 here I am focusing on updating to 3.14.15 and getting
things to the minimal to just compile. Then the idea would be
to fix the possible issues by fixing the metadata.

See that the CSS demo is currently disabled since the CssProvider
is not properly picked up for some reason. We will need to research
a bit on this but for now it is good enough doing it without.

There are also 3 new patches that need to be applied due to the
scanner hanging or giving an error when parsing the source code.
Two of those patches are already in GTK upstream see:
https://git.gnome.org/browse/gtk+/commit/?id=64eca0bb2dad94164d38a89797bf7a439a321b48
https://git.gnome.org/browse/gtk+/commit/?id=fab70d34d443d25c8c68fea30ae7b483749b5e40
The third patch has not yet a workaround that we can merge, it seems
The third patch has not yet a workaround that we can merge, it seems
the ifdef about the GI_SCANNER is giving some issues. We should probably
improve the scanner for this.

See that there are also some generator problems with GtkWidgetClass.
The part of the patch was taken from:
https://github.com/openmedicus/gtk-sharp/commit/5fc657a39cd48ca595992f6bdcbf7bea2c348aed
This workaround fixes the problem below. We should probably research here why this happens and how we can fix it in a better way:
./Widget.cs(111,4): error CS0246: The type or namespace name `GtkWidgetClass' could not be found. Are you missing an assembly reference?
./Widget.cs(112,4): error CS0841: A local variable `klass' cannot be used before it is declared
./Widget.cs(113,4): error CS0103: The name `OverrideClassStruct' does not exist in the current context
./Widget.cs(113,32): error CS0841: A local variable `klass' cannot be used before it is declared

nacho commented:
There we go. Adding also @mikkeljohnsen in case he can give some insight about the GtkWidgetClass issue.
